funderfinder/utils/list_gsoc.py

```diff
@@ -0,0 +1,263 @@
+        for project in year_projects:
+            yield project
+
+
+def get_modern_archive_project(year: int, slug: str) -> dict:
+    """
+    Retrieves project metadata for a project and year after 2016.
+    :param year: Year we want to retreive project metadata from
+    :param slug: The GSOC project slug, retrieved from their API
+    :return: Dict of project metadata
+    """
+    time.sleep(SCRAPE_DELAY)
+    project_url = f"https://summerofcode.withgoogle.com/api/archive/programs/{year}/organizations/{slug}/"
+    meta = requests.get(project_url).json()
+    repos = []
+    repos.extend(get_link_matches(meta["description_html"]))
+    repos.extend(get_link_matches(meta["ideas_list_url"]))
+    for student_project in meta["projects"]:
+        repos.extend(get_link_matches(student_project["abstract_html"]))
+        repos.extend(get_link_matches(student_project["project_code_url"]))
+    return {
+        "name": meta["name"],
+        "link": project_url,
+        "repos": list(set(repos)),
+        "year": year,
+    }
+
+
+def get_modern_archive_projects(year: int) -> tuple:
+    """
+    Retrieves project metadata for a year after 2016, if available
+    :param year: Year to retrieve project metadata from
+    :return: A tuple, with the first element a boolean which is True if we found project metadata for the year and
+    false otherwise (which it may be for the current year), and the second element an iterable of project metadata
+    """
+    org_url = f"https://summerofcode.withgoogle.com/api/archive/programs/{year}/organizations/"
+    orgs = requests.get(org_url).json()
+    is_ok = (type(orgs) == list) and (len(orgs) > 0)
+    if not is_ok:
+        return False, []
+    meta = (get_modern_archive_project(year, org["slug"]) for org in orgs)
+    return True, meta
+
+
+def get_curr_year_project(year: int, slug: str) -> dict:
+    """
+    Get project metadata for a project in the current year
+    :param year: The current year
+    :param slug: GSOC project slug
+    :return: Dict of project metadata
+    """
+    time.sleep(SCRAPE_DELAY)
+    project_url = f"https://summerofcode.withgoogle.com/api/organization/{slug}/"
+    meta = requests.get(project_url).json()
+    repos = []
+    repos.extend(get_link_matches(meta["description"]))
+    repos.extend(get_link_matches(meta["ideas_link"]))
+    repos.extend(get_link_matches(meta["source_code"]))
+    repos.extend(get_link_matches(meta["website_url"]))
+    return {
+        "name": meta["name"],
+        "link": project_url,
+        "repos": list(set(repos)),
+        "year": year,
+    }
+
+
+def get_curr_year_projects(year: int) -> iter:
+    """
+    Get projects for the current year, if available
+    :param year: The current year
+    :return: An generator (possibly empty) of project metadata
+    """
+    org_url = f"https://summerofcode.withgoogle.com/api/program/{year}/organizations/"
+    orgs = requests.get(org_url).json()
+    if type(orgs) == list:
```

Comment:
I think I don't understand this defensive check. Is there a potential type it could otherwise be? Is an empty list a possibility? I think my mind is wondering why a check for type rather than length.

Reply:
The API returns a dict if no data found like so https://summerofcode.withgoogle.com/api/program/2024/organizations/ -- added a comment to make this less confusing!